src/utils/lazy.ts

```diff
@@ -7,21 +7,68 @@ export class Lazy<T> {
     private _isValueCreated: boolean = false;
     private _value: T | undefined;
 
-    public constructor(private readonly valueFactory: () => T) {
+    public constructor(private readonly valueFactory: () => T, private readonly _valueLifetime?: number) {
     }
 
     public get isValueCreated(): boolean {
         return this._isValueCreated;
     }
 
     public get value(): T {
-        if (!this._isValueCreated) {
-            this._value = this.valueFactory();
-            this._isValueCreated = true;
+        if (this._isValueCreated) {
+            return this._value;
+        }
+
+        this._value = this.valueFactory();
+        this._isValueCreated = true;
+
+        if (this._valueLifetime) {
+            const reset = setTimeout(() => {
+                this._isValueCreated = false;
+                this._value = undefined;
+                clearTimeout(reset);
+            }, this._valueLifetime);
         }
 
         return this._value;
     }
 }
 
-export default Lazy;
+export class AsyncLazy<T> {
+    private _isValueCreated: boolean = false;
+    private _value: T | undefined;
+    private _valuePromise: Promise<T> | undefined;
+
+    public constructor(private readonly valueFactory: () => Promise<T>, private readonly _valueLifetime?: number) {
+    }
+
+    public async getValue(): Promise<T> {
+        if (this._isValueCreated) {
+            return this._value;
+        }
+
+        this._isValueCreated = false;
```

Comment:
Yes, so that subsequent calls (while this one is `await`ing) don't think there is a value.

Reply:
My point is on line 46 you check if the value is true and return if it is. So on line 50 it must be false

Reply:
🤦 doh!

I was thinking "I need to set it so that other executions don't think there's a value"...but you're absolutely right, if we got past the above check then it is, by definition, _already_ false.